tempesta_fw/lb/tfw_lb_http.c

```diff
@@ -0,0 +1,1091 @@
+}
+
+static struct sock *
+tfw_lb_sched(TfwLbGroup *group, TfwHttpReq *req)
+{
+	/* Should be done via a function pointer, but KISS for now. */
+	return (group->sched == TFW_LB_SCHED_RR)
+	       ? tfw_lb_sched_rr(group)
+	       : tfw_lb_sched_hash(group, req);
+}
+
+static TfwLbGroup *
+tfw_lb_select_group(TfwHttpReq *req)
+{
+	TfwLbRule *rule;
+
+	if (!tfw_lb_rules)
+		return tfw_lb_default_group;
+
+	rule = tfw_http_match_req_entry(req, tfw_lb_rules, TfwLbRule, rule);
+	return rule ? rule->group : tfw_lb_default_group;
+}
+
+/**
+ * XXX: all useful logic from tfw_sched_http is merged into this function.
+ * The rest of tfw_sched_http is configuration parsing and boilerplate code.
+ */
+static int
+tfw_lb_http_send_msg(TfwMsg *msg)
+{
+	struct sock *sk;
+	TfwLbGroup *group;
+	TfwHttpReq *req = container_of(msg, TfwHttpReq, msg);
+
+	group = tfw_lb_select_group(req);
+	if (unlikely(!group)) {
+		TFW_ERR("can't select a backend_group\n");
+		return -ENOENT;
+	}
+
+	sk = tfw_lb_sched(group, req);
+	if (unlikely(!sk)) {
+		TFW_ERR("can't obtain a connection to the backend_group: %s\n",
+			group->name);
+		return -ENOENT;
+	}
+
+	TFW_DBG("send to backend_group: %s\n", group->name);
+	ss_send(sk, &msg->skb_list);
+	return 0;
+}
+
+/*
+ * ------------------------------------------------------------------------
+ *	Connection management.
+ * ------------------------------------------------------------------------
+ */
+
+#define TFW_LB_CONND_THREAD_NAME "tfw_lb_connd"
+#define TFW_LB_CONND_RETRY_INTEVAL_MSEC 1000
+
+/* The thread that connects to back-end servers in background. */
+static struct task_struct *tfw_lb_connd_task = NULL;
+DECLARE_WAIT_QUEUE_HEAD(tfw_lb_connd_wakeup_wq);
+
+/* Indicates that some connection was closed, so the thread should wake up. */
+static bool tfw_lb_connd_should_wakeup = false;
+
+/**
+ * Re-establish closed connection in background.
+ *
+ * This is a callback which is called when a connection is closed.
+ * Two things should happen then:
+ *  - The connection should be re-established.
+ *  - The dead socket in the TfwLbSrv->conn_pool should be re-placed.
+ *
+ * Instead of doing them here, we defer the job to the tfw_lb_connd thread.
+ * That is done because:
+ *  - The function is called in atomic context, but we don't have a non-blocking
+ *    connect() routine in Synchonous Sockets yet.
+ *  - The function may be run concurrently on multiple CPUs, and a  multi-writer
+ *    implementation for TfwLbSrv->conn_pool is tricky. We can do all writes in
+ *    the single thread and thus simplify the code.
+ */
+static void
+tfw_lb_reconnect_in_bg(struct sock *sk)
+{
+	TFW_DBG("back-end server connection is closed: "
+		"sk=%p sock=%p\n", sk, sk->sk_socket);
+
+	tfw_lb_connd_should_wakeup = true;
+	wake_up(&tfw_lb_connd_wakeup_wq);
+}
+
+/**
+ * Establish a single connection to the back-end server (but don't yet add it
+ * to the server's connection pool).
+ */
+static struct socket *
+tfw_lb_srv_connect(TfwLbSrv *srv)
```

Comment:
The function is mostly copy-paste of tfw_backend_connect() while the logic of the last one is just a special case of tfw_lb_srv_connect(): we can have only one or many connections to one particular server.